protocol/v1/qbft/instance/change_round.go

```diff
@@ -139,53 +146,30 @@ func (i *Instance) actOnExistingPrePrepare(signedMessage *specqbft.SignedMessage
 	return i.UponPrePrepareMsg().Run(msg)
 }
 
-func (i *Instance) changeRoundQuorum(round specqbft.Round) (quorum bool, t int, n int) {
-	// TODO - calculate quorum one way (for prepare, commit, change round and decided) and refactor
-	msgs := i.containersMap[specqbft.RoundChangeMsgType].ReadOnlyMessagesByRound(round)
-	quorum = len(msgs)*3 >= i.ValidatorShare.CommitteeSize()*2
-	return quorum, len(msgs), i.ValidatorShare.CommitteeSize()
+func (i *Instance) changeRoundQuorum(msgs []*specqbft.SignedMessage) (quorum bool, t int, n int) {
+	uniqueSigners := make(map[spectypes.OperatorID]bool)
+	for _, msg := range msgs {
+		for _, signer := range msg.GetSigners() {
+			uniqueSigners[signer] = true
+		}
+	}
+	quorum = len(uniqueSigners)*3 >= i.ValidatorShare.CommitteeSize()*2
+	return quorum, len(uniqueSigners), i.ValidatorShare.CommitteeSize()
 }
 
 func (i *Instance) roundChangeInputValue() ([]byte, error) {
 	// prepare justificationMsg and sig
-	var justificationMsg *specqbft.Message
-	var aggSig []byte
-	var roundChangeJustification []*specqbft.SignedMessage
-	ids := make([]spectypes.OperatorID, 0)
+	data := &specqbft.RoundChangeData{
+		PreparedValue: i.State().GetPreparedValue(),
+		PreparedRound: i.State().GetPreparedRound(),
+	}
 	if i.isPrepared() {
 		quorum, msgs := i.containersMap[specqbft.PrepareMsgType].QuorumAchieved(i.State().GetPreparedRound(), i.State().GetPreparedValue())
 		i.Logger.Debug("change round - checking quorum", zap.Bool("quorum", quorum), zap.Int("msgs", len(msgs)), zap.Any("state", i.State()))
-		var aggregatedSig *bls.Sign
-		justificationMsg = msgs[0].Message
-		for _, msg := range msgs {
-			// add sig to aggregate
-			sig := &bls.Sign{}
-			if err := sig.Deserialize(msg.Signature); err != nil {
-				return nil, err
-			}
-			if aggregatedSig == nil {
-				aggregatedSig = sig
-			} else {
-				aggregatedSig.Add(sig)
-			}
 
-			// add id to list
-			ids = append(ids, msg.GetSigners()...)
-		}
-		aggSig = aggregatedSig.Serialize()
-
-		roundChangeJustification = []*specqbft.SignedMessage{{
-			Signature: aggSig,
-			Signers:   ids,
-			Message:   justificationMsg,
-		}}
+		data.RoundChangeJustification = msgs
```

Comment:
@nkryuchkov spec using array of prepare messages instead of one that holds aggregated msg?